File: SymmetricTree.py
"""
101. Symmetric Tree
Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

For example, this binary tree is symmetric:

    1
   / \
  2   2
 / \ / \
3  4 4  3
But the following is not:
    1
   / \
  2   2
   \   \
   3    3
Note:
Bonus points if you could solve it both recursively and iteratively.
"""
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# O(n)
# O(h)
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        stack = [root.left, root.right]
        while stack:
            p = stack.pop()
            q = stack.pop()
            if p is None and q is None:
                continue
            if p is None or q is None or p.val != q.val:
                return False
            stack.append(p.left)
            stack.append(q.right)
            stack.append(p.right)
            stack.append(q.left)
        return True

class Solution2(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        return self.recur(root.left, root.right)
        
    def recur(self, left, right):
        if left is None and right is None:
            return True
        if left is None or right is None or left.val != right.val:
            return False
        return self.recur(left.left, right.right) and self.recur(left.right, right.left)

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)
    print(Solution().isSymmetric(root))

File: Java/SymmetricTree.java
import java.util.*;
public class SymmetricTree {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) {return true;}
        Stack<TreeNode> left = new Stack<TreeNode>();
        Stack<TreeNode> right = new Stack<TreeNode>();
        left.add(root.left);
        right.add(root.right);
        TreeNode l;
        TreeNode r;
        while(left.size() != 0) {
            l = left.pop();
            r = right.pop();
            if(l == null && r == null) {continue;}
            if(l == null || r == null) {return false;}
            if(l.val != r.val) {return false;}
            left.add(l.left);
            left.add(l.right);
            right.add(r.right);
            right.add(r.left);
        }
        return true;
    }
    
    public boolean isSymmetric2(TreeNode root) {
        if(root == null) {return true;}
        return recur(root.left, root.right);
    }
    
    private boolean recur(TreeNode left, TreeNode right) {
        if(left == null && right == null) {return true;}
        if(left == null || right == null || left.val != right.val) {return false;}
        return recur(left.left, right.right) && recur(left.right, right.left);
    }

}
